### PR TITLE
Partially Fixes #3 : Added image and video subpage

### LINKED ISSUE #3: Create sub pages with different filters
The screenshot below shows different facettes/filters.

screenshot from 2016-10-15 15 44 05

Please implement it in the same style for
- [ ] All https://github.com/fossasia/loklak_search/issues/17
- [ ] Video Results https://github.com/fossasia/loklak_search/issues/49
- [ ] Image Results https://github.com/fossasia/loklak_search/issues/48
- [ ] News Results https://github.com/fossasia/loklak_search/issues/46

### PR DESCRIPTION
**Pull request by Achint08**
**Changes proposed in this pull request**

- Added Images and Videos subpage
- Changed count of results to 100 so that we can get a good number of data with images and videos. 
**Screenshots (if appropriate)** 
screen shot 2017-03-29 at 6 01 09 pm
 
**Problem experienced**
@mariobehling  @daminisatya @sudheesh001 @hemantjadon @uttu357 @rishiraj824 
Hello , the image and video subpage is now partially working.
I am experiencing three problems
First related to iframe . If I am using iframe to display video , I am getting an error which redirects to 
http://g.co/ng/security#xss
Secondly , Server still sends image links with "https:pic.twitter.com/{{id}" so such images are not loaded.
Thirdly, Suggest me some design for tweets with two-three images/videos or tweets with both images and videos
Kindly review and help me in the issues and I will improve it in later commits . 

gh-pages  here](https://achint08.github.io/loklak_search/)

### CODE REVIEW

**Comment by yukiisbored** [comment]
@Achint08 That link shows how to fix the problem, XSS is a pretty serious problem. Since, I'm not that familiar with Angular by any means, I can't give you a suggestion nor a solution.

**Comment by sudheesh001** [comment]
Because of similar issues, the v1 of the webclient used iframely. Do have a look at it https://github.com/itteco/iframely

**Comment by hemantjadon** [comment]
@Achint08 this seems good to me. 
XSS : as mentioned by @sudheesh001 iframely can be used (Though not sure will remove all the errors)

File: src/app/feed/feed.component.ts [hunk 1]
```diff
@@ -243,6 +243,21 @@ export class FeedComponent implements OnInit, OnDestroy {
 		this.store.dispatch(new apiAction.UnSelectLightbox(event));
 	}
 
+	private filterresults(filtervalue) {
+		if(filtervalue == 0) {
+			console.log("All");
```
Comment (uttu357) [inline src/app/feed/feed.component.ts]: Remove console logs

**Comment by uttu357** [comment]
Loading 100 results initially with the images will take a long time and hence increase the loading time of the website considerably... thus instead of getting 100 results initially we can send a request to `"queryString" /image` on clicking on images sub-tab, `"queryString" /video` on video sub-tab and load only 20-30 results initially. This will be better in my opinion.. Your thoughts @hemantjadon @Achint08? Also this will help in pagination for each subtab...

**Comment by Achint08** [comment]
@uttu357 Images will load only when it is in the view port of the device. And thanks for pointing out, we can use same pattern to request for tweets with images and videos. (Y)

File: src/app/feed/feed.component.ts (hunk 1 above)
Comment (Achint08) [inline src/app/feed/feed.component.ts]: Sure. :)

File: src/app/feed/feed-card/feed-card.component.html [hunk 2]
```diff
@@ -26,6 +26,12 @@
 					[text]="itemText"
 					[config]="cardAutolinkerConfig" (click)="showLightBox.emit({show : 'hide'})"></feed-linker>
 			</div>
+			<div class="card-image" *ngFor="let image of feedItem.images let i = index">
```
Comment (rishiraj824) [inline src/app/feed/feed-card/feed-card.component.html]: This looks good to me. Nice Work @Achint08!  Although as @Achint08 pointed out that URLs with pic.twitter.com are still appearing in the JSON, we can either filter those before using feedItem.images in the *ngFor. This can be merged for now and reproduced in a new issue.

**Comment by uttu357** [comment]
screenshot from 2017-03-30 01 22 55
The lightbox design is not responsive and is not able to handle image tweets..

**Comment by rishiraj824** [comment]
@Achint08 @uttu357 Let's work on this issue before merging it in the master branch.

File: src/app/feed/feed-card/feed-card.component.html (hunk 2 above)
Comment (Achint08) [inline src/app/feed/feed-card/feed-card.component.html]: Sure. As soon as I get some of the changes done , I will let other developers work on this issue more and improve it. :)

**Comment by Achint08** [comment]
@uttu357 I know. I will look forward and do all the necessary changes soon. 👍

**Comment by hemantjadon** [comment]
@Achint08 @uttu357 @rishiraj824 
Separating filters from query is i guess necessary. (Separate concerns) FIltering has nothing to do with query being resolved.
The important bit is when all is clicked, the results are fetched, along with images, and when image is clicked the results are already loaded from cache.
What we can do in future is making some kind of component which loads images Lazily and shows it. I will be thinking about this.
For now let's not make this PR complicated, and work on basic error fixing. (Lightbox, console errors, iframely) 
Even i would suggest that make only the reducer for the time being, and how effects are linked to it and actions flow can be taken into account in separate PR's.

**Comment by mariobehling** [comment]
Thank you. Will be great when you get it working.

Please also follow the market leader in the style of the menu bar.

screenshot from 2017-03-30 14-45-31

**Comment by Achint08** [comment]
@mariobehling Sure. Just a day or two. This feature will be working as you say.

**Comment by uttu357** [comment]
@Achint08 When clicking on search more on images tab text results are also shown..

**Comment by hemantjadon** [comment]
@Achint08 as mentioned by @uttu357 the results should be filtered every time with search more and also when some new query is made, will be easier to implement if the separate filter state is used along with separate reducer.

**Comment by mariobehling** [comment]
@Achint08 Thanks for the contribution. Please help to resolve remaining questions and merge conflicts. Cheers.

**Comment by Achint08** [comment]
Sorry for late commits. My mid- semester lab sessions was going on. 
gh-pages [here](https://achint08.github.io/loklak_search/). 

- Videos now working( however it is loading slowly).
- Separate request for image , video tabs.
- Menu bar like market leader. 
However Work on Lightbox UI is left.

File: src/app/feed/feed-header/feed-header.component.scss [hunk 3]
```diff
@@ -10,7 +10,7 @@
 		flex-wrap: wrap;
 		align-items: center;
 		width: 100%;
-		height: 85px;
+		height: 155px;
```
Comment (rishiraj824) [inline src/app/feed/feed-header/feed-header.component.scss]: @Achint08 The menubar height seems changed.

File: src/app/feed/feed-header/feed-header.component.ts [hunk 4]
```diff
@@ -44,5 +45,18 @@ export class FeedHeaderComponent implements OnInit {
...
+			console.log("All");
+		}
+		else if(filtervalue == 1) {
+			console.log("Images");
+		}
+		else if(filtervalue == 2) {
+			console.log("Videos");
+		}
+		this.filtertabs.emit(filtervalue);
+	}
```
Comment (rishiraj824) [inline src/app/feed/feed-header/feed-header.component.ts]: console.logs can be removed.

File: src/app/services/search.service.ts [hunk 5]
```diff
@@ -26,6 +27,7 @@ export class SearchService {
 		searchParams.set('callback', 'JSONP_CALLBACK');
 		searchParams.set('minified', SearchService.minified_results.toString());
 		searchParams.set('source', SearchService.source);
+		searchParams.set('count',SearchService.count.toString());
```
Comment (rishiraj824) [inline src/app/services/search.service.ts]: can the type be 'string' directly in line 40?

File: src/app/feed/feed-card/feed-card.component.ts [hunk 6]
```diff
@@ -126,4 +131,15 @@ export class FeedCardComponent implements OnInit {
 		}
 
 	}
+
+	private sanitizeandembedURLs(links) {
+		for( let i = 0; i < links.length; i++)
+		{
+			let videoid = links[i].match(/(?:https?:\/{2})?(?:w{3}\.)?youtu(?:be)?\.(?:com|be)(?:\/watch\?v=|\/)([^\s&]+)/);
+			if(videoid != null) {
+   				links[i] = "http://www.youtube.com/embed/" + videoid[1];
```
Comment (rishiraj824) [inline src/app/feed/feed-card/feed-card.component.ts]: Is there a way to just show the thumbnail of the video and then load the video in the modal, like twitter does, this can also solve the problem of the site loading slowly.

File: src/app/feed/feed-header/feed-header.component.scss [hunk 7]
```diff
@@ -122,6 +122,11 @@
 		}
 
 	}
+	div.tabs{
```
Comment (rishiraj824) [inline src/app/feed/feed-header/feed-header.component.scss]: tabs can be aligned to the left on mobile.

File: src/app/feed/feed-header/feed-header.component.scss (hunk 3 above)
Comment (Achint08) [inline src/app/feed/feed-header/feed-header.component.scss]: @rishiraj824 To get menubar just like the market leader. Height needs to be changed. 
Any design idea?

File: src/app/feed/feed-card/feed-card.component.ts (hunk 6 above)
Comment (Achint08) [inline src/app/feed/feed-card/feed-card.component.ts]: Will do this. :)

**Comment by Achint08** [comment]
@rishiraj824 Done! (Y)
Checkout [here](https://achint08.github.io/loklak_search/)

File: src/app/feed/feed-header/feed-header.component.scss (hunk 3 above)
Comment (rishiraj824) [inline src/app/feed/feed-header/feed-header.component.scss]: I was talking about the height which is there above the search bar and the logo.

screenshot 97

**Comment by mariobehling** [comment]
Thank you. Please check the following:
* [ ] Avoid round shaped corners
* [ ] Seems like you are changing ratio of image (it is not in a natural dimension)
* [ ] How do you plan to deal with several images? At the moment I see only one image gets loaded at a time. 
screenshot from 2017-04-13 10-10-08
* [ ] More button not working

File: src/app/feed/feed-card/feed-card.component.ts [hunk 8]
```diff
@@ -1,6 +1,7 @@
-import { Component, OnInit, Input, ChangeDetectionStrategy, ChangeDetectorRef, Output, EventEmitter } from '@angular/core';
+import { Component, OnInit , Input, ChangeDetectionStrategy, ChangeDetectorRef, Output, EventEmitter } from '@angular/core';
```
Comment (hemantjadon) [inline src/app/feed/feed-card/feed-card.component.ts]: Unnecessary change

File: src/app/feed/feed-card/feed-card.component.ts [hunk 9]
```diff
@@ -126,4 +131,15 @@ export class FeedCardComponent implements OnInit {
 		}
 
 	}
+
+	private sanitizeandembedURLs(links) {
+		for( let i = 0; i < links.length; i++)
```
Comment (hemantjadon) [inline src/app/feed/feed-card/feed-card.component.ts]: can use more `Array.protoype.forEach(_ => {})`

File: src/app/feed/feed.component.ts [hunk 10]
```diff
@@ -243,6 +243,26 @@ export class FeedComponent implements OnInit, OnDestroy {
 		this.store.dispatch(new apiAction.UnSelectLightbox(event));
 	}
 
+	private filterresults(filtervalue) {
```
Comment (hemantjadon) [inline src/app/feed/feed.component.ts]: Filtering this way, I am not so sure about this. What about filter Actions?

Comment (Achint08) [inline src/app/feed/feed.component.ts]: @hemantjadon I have removed those. 
Previously, we were filtering out feeds with media from all the tweets. So, The number of feeds varied in the media tabs(sometimes 0,1 but not more than 5 most of the time) . Now, we are sending request to the server for tweets with specific media(images/videos) and it is returning those tweets only. This will increase the number of results.
 Moreover, previously, a problem with  pagination was going on. How would we implement pagination for media tabs?

**Comment by Achint08** [comment]
@mariobehling Solved most of the issues. (Y)
Check it out [here](https://achint08.github.io/loklak_search/)
And for multiple images, It will arranged in a table format just like below :-
screen shot 2017-04-14 at 3 18 06 pm
For "Show more" to work, we are not getting feeds from server side from Pagination. Will create an issue there.

**Comment by mariobehling** [comment]
@Achint08 
Top bar height has increased now.
I am unclear questions about two issues has been resolved.

This PR was opened 17 days ago and it tackles a lot of different areas. It looks to me like this PR is too big. It should be possible to solve issues in a perfect world within one day. If not, sub-issues should be opened.

Could you please help and make it easier for reviewers? We need to be able to work faster and in a team. 

as @hemantjadon suggested please reduce the scope of this PR. I suggest to focus on portraying the first image in tweets first and creating an image tab, only showing those tweets with photos.

Honestly, it does not make sense that a lot of people keep reviewing issues over two weeks and longer. Please make commits that can be reviewed more easily and quicker.

Please work as a team and enable other to contribute. One possibility is to create sub-issues and make smaller commits. I hope you understand this and we can close this PR as soon as possible.

**Comment by Achint08** [comment]
@mariobehling I totally understand that and ofcourse, sorry for this long PR. I surely want other contributors to work on this issue. This is a issue with so many sub issues. 
So for now , I am closing this PR and making a new PR with only those changes which I have solved and with explanation. 
Rest I will open issues which reviewers have suggested me and we will go the way contributors want to. :)
